### DISCUSSION TITLE
Override prisma adapter to add new columns

### DISCUSSION THREAD

**Discussion by emulienfou** [nextauthjs/next-auth#2197] [Help]
Hey people, I'm currently using Prisma adapter and want to add a new column for User's `roles`.
I followed [this answer](https://github.com/nextauthjs/next-auth/discussions/805#discussioncomment-351960), however this is working with adapter **TypeORM**.

How can I make it work with Prisma instead?

**Comment by shanewwarren** [comment]
I was struggling figuring out the same thing this weekend.  Here is the approach I've taken:

It seems you can add whatever arbitrary columns to the `User` table as you'd like in `schema.prisma` (https://github.com/nextauthjs/adapters/tree/main/packages/prisma#custom-fields):

**prisma/schema.prisma**
```prisma
model User {
  id            Int       @id @default(autoincrement())
  name          String?
  email         String?   @unique
  emailVerified DateTime? @map(name: "email_verified")
  image         String?
  createdAt     DateTime  @default(now()) @map(name: "created_at")
  updatedAt     DateTime  @updatedAt @map(name: "updated_at")
  role          String    @default("user")
}
```

**pages/api/auth/[...nextauth].ts**
```typescript
  callbacks: {
    async jwt(token, user) {
      if (user?.role) {
        token.role = user.role;
      }

      return token;
    },
    async session(session, token) {
      if (token?.role) {
        session.user.role = token.role;
      }
      return session;
    },
  },
```

Since my users in particular were logging in from Facebook, I wanted to flag and upgrade particular users to the `admin` role.  So I added a callback to the `createUser` event.

**pages/api/auth/[...nextauth].ts**
```typescript
  events: {
    createUser: async (user: User) => {
      try {
        const count = await prisma.admin.count({
          where: {
            email: user.email,
          },
        });

        if (count > 0) {
          await prisma.user.update({
            where: {
              id: user.id,
            },
            data: {
              role: "admin",
            },
          });
        }
      } catch (err) {
        console.log(err);
      }
    },
  },
```

Hope this helps!

**Reply by Standaa** [reply]
The main problem with this approach is that you cannot use it for a field which could be updated like `username`.
If a field value is changed at the DB level, it will require the user to log in again for the changes to be reflected in the token.

**Reply by swport** [reply]
> I was struggling figuring out the same thing this weekend. Here is the approach I've taken:
> 
> It seems you can add whatever arbitrary columns to the `User` table as you'd like in `schema.prisma` (https://github.com/nextauthjs/adapters/tree/main/packages/prisma#custom-fields):
> 
> **prisma/schema.prisma**
> 
> ```prisma
> model User {
>   id            Int       @id @default(autoincrement())
>   name          String?
>   email         String?   @unique
>   emailVerified DateTime? @map(name: "email_verified")
>   image         String?
>   createdAt     DateTime  @default(now()) @map(name: "created_at")
>   updatedAt     DateTime  @updatedAt @map(name: "updated_at")
>   role          String    @default("user")
> }
> ```
> 
> **pages/api/auth/[...nextauth].ts**
> 
> ```ts
>   callbacks: {
>     async jwt(token, user) {
>       if (user?.role) {
>         token.role = user.role;
>       }
> 
>       return token;
>     },
>     async session(session, token) {
>       if (token?.role) {
>         session.user.role = token.role;
>       }
>       return session;
>     },
>   },
> ```
> 
> Since my users in particular were logging in from Facebook, I wanted to flag and upgrade particular users to the `admin` role. So I added a callback to the `createUser` event.
> 
> **pages/api/auth/[...nextauth].ts**
> 
> ```ts
>   events: {
>     createUser: async (user: User) => {
>       try {
>         const count = await prisma.admin.count({
>           where: {
>             email: user.email,
>           },
>         });
> 
>         if (count > 0) {
>           await prisma.user.update({
>             where: {
>               id: user.id,
>             },
>             data: {
>               role: "admin",
>             },
>           });
>         }
>       } catch (err) {
>         console.log(err);
>       }
>     },
>   },
> ```
> 
> Hope this helps!

Wouldn't that be double query to the DB. Overriding PrismaAdapter seems like a better solution.

```javascript
const prismaAdapter = PrismaAdapter(prisma);

// @ts-ignore
prismaAdapter.createUser = (data) => {
  console.log("Data: ", data);
  return prisma.user.create({ data });
};
```
